src/rgw/driver/sfs/sqlite/retry.h

```diff
@@ -0,0 +1,97 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t
+// vim: ts=8 sw=2 smarttab ft=cpp
+/*
+ * Ceph - scalable distributed file system
+ * SFS SAL implementation
+ *
+ * Copyright (C) 2023 SUSE LLC
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation. See file COPYING.
+ */
+#pragma once
+
+#include <functional>
+#include <optional>
+#include <system_error>
+#include <thread>
+
+#include "errors.h"
+#include "rgw_perf_counters.h"
+
+using namespace std::chrono_literals;
+namespace rgw::sal::sfs::sqlite {
+
+/// RetrySQLite is a utility to retry non-critically failed
+// sqlite_orm. run() executes function passed with the constructor
+// returning the result OR an empty optional if non could be obtained
+// after retrying. Catches all non-critical exceptions and makes them
+// available via failed_error(). Critical exceptions are passed on.
+template <typename Return>
+class RetrySQLite {
+ public:
+  using Func = std::function<Return(void)>;
+
+ private:
+  const Func m_fn;
+  const int m_max_retries{10};
+  bool m_successful{false};
+  int m_retries{0};
+  std::error_code m_failed_error{};
+
+ public:
+  RetrySQLite(Func&& fn) : m_fn(std::forward<Func>(fn)) {}
+  RetrySQLite(RetrySQLite&&) = delete;
+  RetrySQLite(const RetrySQLite&) = delete;
+  RetrySQLite& operator=(const RetrySQLite&) = delete;
+  RetrySQLite& operator=(RetrySQLite&&) = delete;
```

Comment:
I totally like your approach! 

But, I don't think we are gonna need it :( 

We only need to retry transactions. Possibly only the read then write kind. Everything else is covered by the sqlite retry handler already. 

So far I found about 3-5 instances where we need transaction retries. In general we should use this as seldom as possible to not mask problems. In those places rewriting is necessary anyway to get to a coherent error handling story. 

I think we should do most of the error handling in the SQLite... functions (see the recent ADR). This would push it to the layer above (types.h or SAL) which IMHO should not care about retries, just if a thing happend or not.